// mocha-with-coverage/src/index.js
'use strict';

const gulp = require('gulp');
const mocha = require('gulp-mocha');
const istanbul = require('gulp-istanbul');

module.exports = function (argv) {
    gulp.task('prepare', function () {
        return gulp.src(argv.sources)
            .pipe(istanbul(argv.istanbul))
            .pipe(istanbul.hookRequire());
    });

    gulp.task('run', ['prepare'], function () {
        return gulp.src(argv.tests)
            .pipe(mocha(argv.mocha))
            .pipe(istanbul.writeReports({
                reporters: ['json', 'lcovonly', 'html', 'cobertura'],
                reportOpts: {
                    json: {
                        dir: './reports/coverage/json'
                    },
                    lcovonly: {
                        dir: './reports/coverage/lcovonly'
                    },
                    html: {
                        dir: './reports/coverage/html'
                    },
                    cobertura: {
                        dir: './reports/coverage/cobertura'
                    }
                }
            }));
    });

    gulp.start('run');
};
